File: C/Searching/Searching_03.c
/*
===============================================================================
Program Name    : Linear Search
Program Number  : 03
File Name       : Searching_03.c
Description     : This program implements Linear Search to search for a
                  particular element in a dynamically allocated array.

Date            : August 2026
Language        : C
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Pointer to Structure
3. Typedef
4. Dynamic Memory Allocation
5. Linear Search
6. Array Traversal

Key Learning Points:
- Understand the concept of Linear Search.
- Search an element sequentially from the beginning of the array.
- Return the result based on whether the element is present or absent.
- Analyze the time complexity of Linear Search.

===============================================================================
*/

#include<stdio.h>
#include<stdlib.h>

struct Searching
{
    int *Arr;
    int iSize;
};

typedef struct Searching SEARCHING;
typedef struct Searching *PSEARCHING;

void Accept(PSEARCHING first)
{
    int i = 0;

    printf("Enter the Elements:\n");

    for(i = 0; i < first->iSize; i++)
    {
        scanf("%d",&first->Arr[i]);
    }
}

void Display(PSEARCHING first)
{
    int i = 0;

    printf("Elements of the Array are:\n");

    for(i = 0; i < first->iSize; i++)
    {
        printf("%d\n",first->Arr[i]);
    }
}

int LinearSearch(PSEARCHING first, int iNo)
{
    int i = 0;

    for(i = 0; i < first->iSize; i++)
    {
        if(first->Arr[i] == iNo)
        {
            return 1;
        }
    }

    return 0;
}

int main()
{
    SEARCHING sobj;
    int iRet = 0;

    sobj.iSize = 5;

    sobj.Arr = (int *)malloc(sizeof(int) * sobj.iSize);

    Accept(&sobj);
    Display(&sobj);

    iRet = LinearSearch(&sobj,30);

    if(iRet == 1)
    {
        printf("Element is Present\n");
    }
    else
    {
        printf("There is no such element\n");
    }

    free(sobj.Arr);

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Enter the Elements:
10
20
30
40
50

Elements of the Array are:
10
20
30
40
50

Element is Present

===============================================================================
Memory Representation:
===============================================================================

              sobj
        +------------------+
        | Arr ------------ |--------+
        | iSize = 5        |        |
        +------------------+        |
                                    v
                         +------+------+------+------+------+
                         |  10  |  20  |  30  |  40  |  50  |
                         +------+------+------+------+------+
                           Arr[0] Arr[1] Arr[2] Arr[3] Arr[4]

Linear Search:

10  ->  Not Match
20  ->  Not Match
30  ->  Match Found

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Define the Searching structure.
3. Create a SEARCHING object.
4. Set the array size.
5. Allocate memory dynamically for the array.
6. Accept the array elements from the user.
7. Display the array elements.
8. Take the element to be searched.
9. Start traversing the array from the first element.
10. Compare each array element with the search element.
11. If a match is found, return 1.
12. If the complete array is searched without a match, return 0.
13. Display whether the element is present or absent.
14. Release the allocated memory using free().
15. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
   Create SEARCHING Object
          |
          v
    Allocate Array Memory
          |
          v
        Accept()
          |
          v
       Display()
          |
          v
    LinearSearch(30)
          |
          v
   Compare Each Element
          |
       +--+--+
       |     |
     Match  No Match
       |     |
       v     v
   Return 1  Continue
       |     |
       |    Complete Array?
       |     |
       |     v
       |   Return 0
       |     |
       +--+--+
          |
          v
    Display Result
          |
          v
        free()
          |
          v
         Stop

===============================================================================
Time Complexity:
===============================================================================

Best Case    : O(1)   → Element found at the first position
Average Case : O(n)
Worst Case   : O(n)   → Element found at the last position or absent

Space Complexity : O(n)

Where n is the size of the dynamically allocated array.

===============================================================================
Conclusion:
===============================================================================

Program 003 implements Linear Search in C using a dynamically allocated
array. The program checks each element sequentially until the required
element is found or the complete array is traversed. Linear Search is
simple and works on both sorted and unsorted arrays, with O(n) worst-case
time complexity.

===============================================================================
*/

File: Java/Searching/Searching_03.java
/*
===============================================================================
Program Name    : Searching - Display Elements
Program Number  : 03
File Name       : Searching_03.java

Description     : This program accepts integer elements from the user and
                  displays the elements stored in the array.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Constructor
4. Array
5. Scanner Class
6. User Input
7. Accept Operation
8. Display Operation
9. Array Traversal
10. Private Data Members

Key Learning Points:
- Understand how to display elements of an array.
- Learn how to traverse an array using a for loop.
- Understand the Accept and Display operations.
- Learn how to access array elements through class methods.
- Understand array indexing in Java.

===============================================================================
*/

import java.util.*;

class Searching
{
    private int Arr[];
    private int iSize;

    public Searching(int iNo)
    {
        iSize = iNo;
        Arr = new int[iSize];
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Accept
    // Description   : Accept elements from user
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Accept()
    {
        Scanner sobj = new Scanner(System.in);

        int i = 0;

        System.out.println("Enter the Elements : ");

        for(i = 0; i < iSize; i++)
        {
            Arr[i] = sobj.nextInt();
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Display
    // Description   : Display elements of the array
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Display()
    {
        int i = 0;

        System.out.println("\nElements of the Array are : ");

        for(i = 0; i < iSize; i++)
        {
            System.out.println(Arr[i]);
        }
    }
}

class Searching_03
{
    public static void main(String A[])
    {
        Searching sobj = new Searching(5);

        sobj.Accept();
        sobj.Display();
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Enter the Elements :
11
21
51
101
151

Elements of the Array are :
11
21
51
101
151

===============================================================================
Memory Representation:
===============================================================================

Searching Object

       sobj
        |
        v
+----------------------+
| Searching Object     |
+----------------------+
| iSize = 5            |
| Arr ----------------------+
+----------------------+     |
                             v
                       +----+----+----+-----+-----+
                       | 11 | 21 | 51 | 101 | 151 |
                       +----+----+----+-----+-----+
                         0    1    2     3     4

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create the Searching class.
3. Declare an integer array Arr and integer iSize.
4. Create a Searching object with size 5.
5. Call the Accept() function.
6. Accept 5 integer elements from the user.
7. Store the elements in the array.
8. Call the Display() function.
9. Traverse the array using a for loop.
10. Display each element of the array.
11. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
   Create Searching Object
          |
          v
      Call Accept()
          |
          v
    Accept Array Elements
          |
          v
     Store Elements
          |
          v
     Call Display()
          |
          v
    Traverse the Array
          |
          v
    Display Elements
          |
          v
         Stop

===============================================================================
Time Complexity:
===============================================================================

Accept()  : O(n)

Display() : O(n)

Overall   : O(n)

Where n is the number of elements in the array.

===============================================================================
Space Complexity:
===============================================================================

O(n) for storing n integer elements in the array.

===============================================================================
Conclusion:
===============================================================================

Program 03 demonstrates how to accept and display elements of an array using
a Searching class in Java. It demonstrates array traversal, user input and
the use of separate Accept() and Display()
